core/unix/coredump.c

```diff
@@ -0,0 +1,362 @@
+                    strhash_hash_add(GLOBAL_DCONTEXT, string_htable, iter.comment,
+                                     (void *)string_table_offset);
+                    offset = string_table_offset;
+                    const size_t comment_len = d_r_strlen(iter.comment);
+                    if (comment_len + string_table_offset >
+                        MAX_SECTION_NAME_BUFFER_SIZE) {
+                        SYSLOG_INTERNAL_ERROR("Section name table is too small to store "
+                                              "all the section names.");
+                        return false;
+                    }
+                    d_r_strncpy(&string_table[string_table_offset], iter.comment,
+                                comment_len);
+                    string_table_offset += comment_len + 1;
+                    string_table[string_table_offset - 1] = '\0';
+                }
+            }
+            seciion_data_size += iter.vm_end - iter.vm_start;
+            ++section_count;
+        }
+        // Add the string table section. Append the section name ".shstrtab" to the
+        // section names table.
+        const int section_header_table_len = d_r_strlen(SECTION_HEADER_TABLE) + 1;
+        d_r_strncpy(&string_table[string_table_offset], SECTION_HEADER_TABLE,
+                    section_header_table_len);
+        string_table_offset += section_header_table_len;
+        ++section_count;
+        seciion_data_size += string_table_offset;
+        memquery_iterator_stop(&iter);
+    }
+
+    file_t elf_file;
+    char dump_core_file_name[MAXIMUM_PATH];
+    if (!get_unique_logfile(".elf", dump_core_file_name, sizeof(dump_core_file_name),
+                            false, &elf_file) ||
+        elf_file == INVALID_FILE) {
+        SYSLOG_INTERNAL_ERROR("Unable to open the core dump file.");
+        return false;
+    }
+
+    if (!write_elf_header(elf_file, /*entry_point=*/0,
+                          /*section_header_table_offset*/ sizeof(ELF_HEADER_TYPE) +
+                              1 /*program_header_count*/ *
+                                  sizeof(ELF_PROGRAM_HEADER_TYPE) +
+                              seciion_data_size,
+                          /*flags=*/0,
+                          /*program_header_count=*/1,
+                          /*section_header_count=*/section_count,
+                          /*section_string_table_index=*/section_count - 1)) {
+        os_close(elf_file);
+        return false;
+    }
+    // TODO i#7046: Fill the program header with valid data.
+    if (!write_phdrs(elf_file, PT_NULL, PF_X, /*offset=*/0, /*virtual_address=*/0,
+                     /*physical_address=*/0,
+                     /*file_size=*/0, /*memory_size=*/0, /*alignment=*/0)) {
+        os_close(elf_file);
+        return false;
+    }
+    int total = 0;
+    if (memquery_iterator_start(&iter, NULL, /*may_alloc=*/false)) {
+        while (memquery_iterator_next(&iter)) {
```

Comment:
Is there a race where the memory regions could change between these 3 iterations?  Are all other threads suspended?  Is it better to have the 1st walk record the bounds and then the subsequent walks use that info instead of another live iteration?

Reply:
I've updated to code to combine all three loops into one and use an array to save data needed later. PTAL.